File: Python/DesignPattern/CreationalDesignPattern/AdapterDesignPattern.py
from abc import ABC, abstractmethod


class IPaymentProcessor(ABC):
    @abstractmethod
    def pay(self): pass


class InHousePaymentProcessor(IPaymentProcessor):
    def pay(self):
        print("In House Application is processing the payment")
        # 100000 lines of codes


class Razorpay:
    def make_payment(self):
        print("Razorpay Payment 10203013203103")


class RazorPayAdapter(IPaymentProcessor):
    def __init__(self):
        self._razorpay = Razorpay()

    def pay(self):
        self._razorpay.make_payment()


class Application:
    def __init__(self, processor: IPaymentProcessor):
        self._payment_processor = processor

    def process_payment(self):
        self._payment_processor.pay()


if __name__ == "__main__":
    app = Application(RazorPayAdapter())
    app.process_payment()

File: C++/DesignPattern/CreationalDesignPattern/AdapterDesignPattern.cpp
#include <iostream>
#include <memory>
using namespace std;

struct IPaymentProcessor {
    virtual void pay() = 0;
    virtual ~IPaymentProcessor() = default;
};

struct InHousePaymentProcessor : IPaymentProcessor {
    void pay() override {
        cout << "In House Application is processing the payment\n";
        // 100000 lines of codes
    }
};

struct Razorpay {
    void make_payment() {
        cout << "Razorpay Payment 10203013203103\n";
    }
};

struct RazorPayAdapter : IPaymentProcessor {
    RazorPayAdapter() : razorpay(make_unique<Razorpay>()) {}

    void pay() override {
        razorpay->make_payment();
    }
private:
    unique_ptr<Razorpay> razorpay;
};

class Application {
    unique_ptr<IPaymentProcessor> paymentProcessor;
public:
    explicit Application(unique_ptr<IPaymentProcessor> processor)
        : paymentProcessor(move(processor)) {}

    void processPayment() {
        paymentProcessor->pay();
    }
};

int main() {
    Application app(make_unique<RazorPayAdapter>());
    app.processPayment();
    return 0;
}
